src/prototype.rs

```diff
@@ -1546,51 +1594,223 @@ fn make_code_span(tree: &mut Tree<Item>, s: &str, open: usize, close: usize) {
+                if let Some(c) = scanner.next_char() {
+                    if !(c <= '\x7f' && is_ascii_punctuation(c as u8)) {
+                        url.push('\\');
+                    }
+                    url.push(c);
+                } else {
+                    return None;
+                }
+            }
+            _ => url.push(c),
+        }
+    }
+    None
+}
+
+fn scan_link_destination_pointy(scanner: &mut InlineScanner) -> Option<String> {
+    if !scanner.scan_ch(b'<') {
+        return None;
+    }
+    let mut url = String::new();
+    while let Some(c) = scanner.next_char() {
+        match c {
+            '>' => return Some(url),
+            '\x00'..='\x1f' | '<' => return None,
+            '\\' => {
+                let c = scanner.next_char()?;
+                if !(c <= '\x7f' && is_ascii_punctuation(c as u8)) {
+                    url.push('\\');
+                }
+                url.push(c);
+            }
+            _ => url.push(c),
+        }
+    }
+    None
+}
+
+fn scan_link_destination(scanner: &mut InlineScanner) -> Option<String> {
+    let save = scanner.clone();
+    if let Some(url) = scan_link_destination_pointy(scanner) {
+        return Some(url);
+    }
+    *scanner = save;
+    scan_link_destination_plain(scanner)
+}
+
+fn scan_link_title(scanner: &mut InlineScanner) -> Option<String> {
+    let open = scanner.next_char()?;
+    if !(open == '\'' || open == '\"' || open == '(') {
+        return None;
+    }
+    let mut title = String::new();
+    let mut nest = 0;
+    while let Some(c) = scanner.next_char() {
+        if c == open {
+            if open == '(' {
+                nest += 1;
+            } else {
+                return Some(title);
+            }
+        }
+        if open == '(' && c == ')' {
+            if nest == 0 {
+                return Some(title);
+            } else {
+                nest -= 1;
+            }
+        }
+        if c == '\\' {
+            let c = scanner.next_char()?;
+            if !(c <= '\x7f' && is_ascii_punctuation(c as u8)) {
+                title.push('\\');
+            }
+            title.push(c);
+        } else {
+            title.push(c);
+        }
+    }
+    None
+}
+
+fn scan_inline_link(scanner: &mut InlineScanner) -> Option<(String, String)> {
+    if !scanner.scan_ch(b'(') {
+        return None;
+    }
+    scanner.scan_while(is_ascii_whitespace);
+    let url = scan_link_destination(scanner)?;
+    let mut title = String::new();
+    let save = scanner.clone();
+    if scanner.scan_while(is_ascii_whitespace) > 0 {
+        if let Some(t) = scan_link_title(scanner) {
+            title = t;
+            scanner.scan_while(is_ascii_whitespace);
+        } else {
+            *scanner = save;
+        }
+    }
+    if !scanner.scan_ch(b')') {
+        return None;
+    }
+    Some((url, title))
+}
+
+// TODO: if it's just node, get rid of struct. But I think there probably will be
+// state related to normalization.
```

Comment:
What kind of normalization?

Reply:
The [label matching algorithm](https://spec.commonmark.org/0.28/#matches). My thinking here is that we compute the normalized form of the link label at the time that it's scanned here, then refer to that normalized string later when determining matches. But maybe that can be done as a normalized-compare.